### PR TITLE
#790 Add Ops\Boxes

### PR DESCRIPTION
**Pull request by xhuan8**
Add nms method from torch vision.
Build torchvision dll from source and rename to "libtorchvision", put it and all dependencies to the same path of exe file.
Tested on windows, the dll uploaded to https://drive.google.com/drive/folders/14R6_HzDEOlespcOd6pAWYgxjoYjD1Xpz?usp=share_link

command to build torch vision:
```
cmake .. -DWITH_CUDA=on -DCMAKE_PREFIX_PATH="D:\Code\libtorch-cuda-11.7\libtorch;D:\Lib\vcpkg\packages\zlib_x64-windows;D:\Lib\vcpkg\packages\libpng_x64-windows;D:\Lib\libjpeg-turbo64\lib" -G "Visual Studio 16 2019"
```

### CODE REVIEW

File: src/Native/LibTorchSharp/Utils.h [hunk 1]
```diff
@@ -296,4 +309,46 @@ torch::nn::init::NonlinearityType get_nl_type(const int64_t nl)
...
+        char buff[FILENAME_MAX];
+        GetCurrentDir(buff, FILENAME_MAX);
+        std::string current_working_dir(buff);
+
+        torch_last_err = strdup(("Failed to load library: " + libName + " " +
+            std::to_string(GetLastError()) + " " + current_working_dir).c_str());
+        return NULL;
+}
+#else
+    lib = dlopen((libName + ".so").c_str(), RTLD_LAZY);
```
Comment (NiklasGustafsson) [inline src/Native/LibTorchSharp/Utils.h:337]: This won't work on Mac, I believe. There, it's '.dylib'

File: src/TorchSharp/Tensor/Tensor.Math.cs [hunk 2]
```diff
@@ -2754,7 +2754,14 @@ public static Tensor einsum(string equation, params Tensor[] tensors)
         /// </summary>
         /// <param name="input">The first input tensor</param>
         /// <param name="other">The second input tensor</param>
-        public static Tensor maximum(Tensor input, Tensor other) => input.maximum(other);
+        public static Tensor max(Tensor input, Tensor other) => maximum(input, other);
```
Comment (NiklasGustafsson) [inline src/TorchSharp/Tensor/Tensor.Math.cs:2757]: I don't think 'max' and 'maximum' are aliases of each other.

https://pytorch.org/docs/stable/generated/torch.max.html?highlight=max#torch.max
https://pytorch.org/docs/stable/generated/torch.maximum.html?highlight=maximum#torch.maximum

File: src/TorchSharp/Tensor/Tensor.Math.cs [hunk 3]
```diff
@@ -2797,7 +2804,14 @@ public static Tensor einsum(string equation, params Tensor[] tensors)
         /// </summary>
         /// <param name="input">The first input tensor</param>
         /// <param name="other">The second input tensor</param>
-        public static Tensor minimum(Tensor input, Tensor other) => input.minimum(other);
+        public static Tensor min(Tensor input, Tensor other) => minimum(input, other);
```
Comment (NiklasGustafsson) [inline src/TorchSharp/Tensor/Tensor.Math.cs:2807]: See comment on 'max'

File: src/TorchSharp/Utils/DeconstructExtension.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+using System.Text;
+
+namespace TorchSharp.Utils
+{
+    /// <summary>
+    /// Converts IEnumerable to tuple.
+    /// </summary>
+    public static class DeconstructExtension
+    {
+        public static void Deconstruct<T>(this IEnumerable<T> seq, out T first, out IEnumerable<T> rest)
```
Comment (NiklasGustafsson) [inline src/TorchSharp/Utils/DeconstructExtension.cs:13]: Doc comments on all public APIs, please.

File: src/TorchVision/Ops/Boxes.cs [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+using System.Collections.Generic;
+
+#nullable enable
+namespace TorchSharp
+{
+    public static partial class torchvision
+    {
+        public static partial class ops
+        {
+            [DllImport("LibTorchSharp")]
```
Comment (NiklasGustafsson) [inline src/TorchVision/Ops/Boxes.cs:33]: imported methods have now been moved to separate files.

**Review by NiklasGustafsson — CHANGES_REQUESTED** [review]

File: test/TorchSharpTest/TestTorchVisionOps.cs [hunk 6]
```diff
@@ -20,7 +20,7 @@ public void NMS_OneBox()
                 });
                 var scores = torch.from_array(new[] { 0.9 });
 
-                var nms_boxes = nms(boxes, scores);
+                var nms_boxes = nms_custom(boxes, scores);
```
Comment (NiklasGustafsson) [inline test/TorchSharpTest/TestTorchVisionOps.cs:23]: This means there's no unit test for 'nms' -- please add one.

**Comment by NiklasGustafsson** [comment]
@huangyangyu -- can you take a look at this PR? You added the C# implementation of 'nms'

**Comment by NiklasGustafsson** [comment]
I'm curious about libtorchvision.dll -- you uploaded it to Google drive, but we'll have to come up with a more automated way of building it.

**Comment by NiklasGustafsson** [comment]
I'm wondering if it wouldn't be more convenient, in the long run to just add torchvision (and audio, and text) as linked projects and build the native code into our native project.

File: src/TorchSharp/Tensor/Tensor.Math.cs (hunk 2 above)
Comment (xhuan8) [reply src/TorchSharp/Tensor/Tensor.Math.cs:2757]: there is an overload at the bottom of the 'max‘ page

torch.max(input, other, *, out=None) → [Tensor](https://pytorch.org/docs/stable/tensors.html#torch.Tensor)
See [torch.maximum()](https://pytorch.org/docs/stable/generated/torch.maximum.html#torch.maximum).

**Comment by xhuan8** [comment]
@NiklasGustafsson  it will be great if it's possible to build torchvision automatically, but currently I cannot find a way to do it.

**Comment by xhuan8** [comment]
Currently my branch is a mess, I'll consider create pull request later.

**Pull request closed without merging**